src/OpenTelemetry.Adapter.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,115 @@
+﻿// <copyright file="GrpcClientDiagnosticListener.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Adapter.Dependencies.Implementation
+{
+    internal class GrpcClientDiagnosticListener : ListenerHandler
+    {
+        private readonly PropertyFetcher startRequestFetcher = new PropertyFetcher("Request");
+
+        public GrpcClientDiagnosticListener(Tracer tracer)
+            : base("Grpc.Net.Client", tracer)
+        {
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStartActivity";
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            {
+                AdapterEventSource.Log.NullPayload(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            if (request.Headers.Contains("traceparent"))
+            {
+                // this request is already instrumented, we should back off
+                return;
+            }
+
+            // The path is also available from the grpc.method tag on the activity. Maybe that should be used instead
+            var path = request.RequestUri.AbsolutePath.Trim('/');
+            this.Tracer.StartActiveSpanFromActivity(path, activity, SpanKind.Client, out var span);
+
+            if (span.IsRecording)
+            {
+                var rx = new Regex(@"(?<package>\w+).(?<service>\w+)/(?<method>\w+)", RegexOptions.Compiled);
```

Comment:
This `Regex` should a read-only static field. Moreover, does it really need to be a regex?

Reply:
Yea agreed it may not need to be a regex. I'm actually not quite sure this is right anyways. There is some discussion of possible prefixes in this path: https://github.com/grpc/grpc-dotnet/issues/110.